py/server/tests/test_table_factory.py

```diff
@@ -478,6 +475,48 @@ def test_j_input_wrapping(self):
         self.assertFalse(isinstance(t, InputTable))
         self.assertTrue(isinstance(t, Table))
 
+    def test_input_table_async(self):
+        cols = [
+            bool_col(name="Boolean", data=[True, False]),
+            byte_col(name="Byte", data=(1, -1)),
+            char_col(name="Char", data='-1'),
+            short_col(name="Short", data=[1, -1]),
+            int_col(name="Int", data=[1, -1]),
+            long_col(name="Long", data=[1, -1]),
+            long_col(name="NPLong", data=np.array([1, -1], dtype=np.int8)),
+            float_col(name="Float", data=[1.01, -1.01]),
+            double_col(name="Double", data=[1.01, -1.01]),
+            string_col(name="String", data=["foo", "bar"]),
+        ]
+        t = new_table(cols=cols)
+
+        with self.subTest("async add"):
+            self.assertEqual(t.size, 2)
+            success_count = 0
+            def on_success():
+                nonlocal success_count
+                success_count += 1
+            append_only_input_table = input_table(col_defs=t.definition)
+            append_only_input_table.add_async(t, on_success=on_success)
+            append_only_input_table.add_async(t, on_success=on_success)
+            while success_count < 2:
+                sleep(0.1)
```

Comment:
@rcaudy is going to cry

Reply:
Unfortunately this is the best I can do though I think it is safe/deterministic . Initially I used `await_update` but because it is what `add_async` uses behind the scene to wait for UGP to finish processing, it creates a race condition where one of the calls will wait for ever.